### PR TITLE
Implement preliminary creation of Wasm GC types

### PR DESCRIPTION
**Pull request by dcodeIO**
Adds logic to incrementally create Wasm GC types with Binaryen's type builder. Is necessary because Binaryen is switching to utilize GC types internally to have all the information, even when the GC feature is disabled, then emitting reference types. Required by #2531 so we can create `ref.func` expressions with concrete types (cc @MaxGraey).

Still some TODOs:

* There is no impl for GC `array` members yet. An `Array<T>` for example would wrap a GC `array`, while an `ArrayBuffer` would be represented by one. However, since the respective code and stdlib changes have not yet been made, yet would go in tandem with the types, this is left out for now. My expectation is that we'd add a source-level (lower case) `array<T>` type that is directly bound to the respective Wasm GC type for stdlib to use, perhaps somewhat like:
  ```ts
  class Array<T> {
    buffer: array<T>;
    ...
  }
  class StaticArray<T> extends array<T> {
    ...
  }
  class ArrayBuffer extends array<u8> {
    ...
  }
  ```
* GC type creation has been tested against the full test suite, being able to represent all the types we have as GC types (with the above limitation), except for one test: duplicate-fields. This test includes a class structure akin to the following:
  ```ts
  class Foo {}
  class Bar extends Foo {}
  class A {
    foo: Foo;
  }
  class B extends A {
    foo: Bar;
  }
  ```
  where there is a duplicate field `foo` with a subtyping relationship. <del>This is sound afaict, but</del> appears to be [not supported by Wasm GC](https://github.com/WebAssembly/binaryen/blob/3e44a9fc542572592ee1ee1bdec19bf1904faaa3/src/wasm/wasm-type.cpp#L1827-L1828), unless both `foo`s would be immutable fields, which they aren't in our case. Looks like we'd have to revert support for specializing duplicate fields like this in order to use Wasm GC (see below).
* Building GC types on the side has side effects (on the text format), in that in order to create the types, a type builder must be created, which requires setting a type system, and then leads to Wasm GC text format with `func_subtype` and whatnot, even though the GC feature is disabled. See: https://github.com/WebAssembly/binaryen/issues/5136#issuecomment-1279838333
* Regarding #2531, what this allows us to do is to `ensureType(theFunctionSignature)` for use with `Module#ref_func` to make it work again, even if the GC type isn't otherwise used and Binaryen would output `funcref` iiuc.

- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

**Review by MaxGraey — APPROVED** [review]
LGTM

**Comment by dcodeIO** [comment]
Reading the GC explainer wrt. the failing duplicate-fields test, the rules in play are

> Structure types support width and depth subtyping
> 
> * `struct <fieldtype1>* <fieldtype1'>* <: struct <fieldtype2>*`
>   * iff `(<fieldtype1> <: <fieldtype2>)*`

> Field types are covariant if they are immutable, invariant otherwise
> 
> * `const <valtype1> <: const <valtype2>`
>   * iff `<valtype1> <: <valtype2>`
> * `var <valtype> <: var <valtype>`
> * Note: mutable fields are not subtypes of immutable ones, so const really means constant, not read-only

i.e. mutable fields are invariant. I guess there is a reason for this limitation, though I don't currently see it.

File: src/module.ts [hunk 1]
```diff
@@ -63,30 +63,28 @@ export namespace TypeRef {
   // special types
   export const None: TypeRef = 0 /* _BinaryenTypeNone */;
   export const Unreachable: TypeRef = 1 /* _BinaryenTypeUnreachable */;
+  export const Auto: TypeRef = -1 /* _BinaryenTypeAuto */;
```
Comment (MaxGraey) [inline src/module.ts:66]:
```suggestion
  export const Auto: TypeRef = -1 /* _BinaryenTypeAuto */;
  export const None: TypeRef = 0 /* _BinaryenTypeNone */;
  export const Unreachable: TypeRef = 1 /* _BinaryenTypeUnreachable */;
```

Comment (dcodeIO) [reply src/module.ts]: Figured that we are not really using the `Auto` type, and given that it's always better to be explicit, I removed it.

**Pull request merged**